src/visualization/makie_visualization.jl

```diff
@@ -0,0 +1,373 @@
+  xfp, yfp = map(reshape_and_interpolate, (x, y))
+  ufp = StructArray{SVector{nvars, uEltype}}(map(reshape_and_interpolate, StructArrays.components(u)))
+
+  # convert variables based on solution_variables mapping
+  solution_variables_ = digest_solution_variables(equations, solution_variables)
+  variable_names = SVector(varnames(solution_variables_, equations))
+
+  transform_to_solution_variables!(uplot, solution_variables_, equations)
+  transform_to_solution_variables!(ufp, solution_variables_, equations)
+
+  return UnstructuredPlotData2D(xplot, yplot, uplot, t, xfp, yfp, ufp, variable_names)
+end
+
+function generate_plotting_triangulation(pds::PlotDataSeries2D{<:UnstructuredPlotData2D};
+                                         set_z_coordinate_zero = false)
+
+  @unpack variable_id = pds
+  pd = pds.plot_data
+  @unpack x, y, u, t = pd
+
+  makie_triangles = Makie.to_triangles(t)
+
+  # trimesh[i] holds GeometryBasics.Mesh containing plotting information on the ith element.
+  num_plotting_nodes, num_elements = size(x)
+  trimesh = Vector{GeometryBasics.Mesh{3, Float32}}(undef, num_elements)
+  coordinates = zeros(Float32, num_plotting_nodes, 3)
+  for element in Base.OneTo(num_elements)
+    for i in Base.OneTo(num_plotting_nodes)
+      coordinates[i, 1] = x[i, element]
+      coordinates[i, 2] = y[i, element]
+      if set_z_coordinate_zero == false
+        coordinates[i, 3] = u[i, element][variable_id]
+      end
+    end
+    trimesh[element] = GeometryBasics.normal_mesh(Makie.to_vertices(coordinates), makie_triangles)
+  end
+  plotting_mesh = merge([trimesh...]) # merge meshes on each element into one large mesh
+  return plotting_mesh
+end
+
+function generate_plotting_wireframe(pds::PlotDataSeries2D{<:UnstructuredPlotData2D};
+                                     set_z_coordinate_zero = false)
+  @unpack variable_id = pds
+  pd = pds.plot_data
+  @unpack xf, yf, uf = pd
+
+  if set_z_coordinate_zero==true
+    sol_f = zeros(eltype(first(uf)), size(xf)) # plot 2d surface by setting z coordinate to zero
+  else
+    sol_f = StructArrays.component(uf, variable_id)
+  end
+  xyz_wireframe = Makie.Point.(map(x->vec(vcat(x, fill(NaN, 1, size(x, 2)))), (xf, yf, sol_f))...)
+  return xyz_wireframe
+end
+
+"""
+    trixi_plot(sol::TrixiODESolution;
+               solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
+
+Creates an interactive surface plot of the solution and mesh for an `UnstructuredMesh2D` type.
+
+Inputs:
+- solution_variables: either `nothing` or a variable transformation function (e.g., `cons2prim`)
+- nvisnodes: number of visualization nodes per dimension
+"""
+function trixi_plot(sol::TrixiODESolution;
+                    solution_variables=nothing, nvisnodes=5, variable_to_plot_in = 1)
+
+  pd = PlotData2D(sol; solution_variables=solution_variables, nvisnodes=nvisnodes)
+
+  @unpack variable_names = pd
+
+  semi = sol.prob.p
+  @unpack mesh = semi
```

Comment:
IIRC, one is supposed to the MESC extraction function for this purpose:
```suggestion
  @unpack variable_names = pd
  mesh, _, _, _ = mesh_equations_solver_cache(sol.prob.p)
```

Reply:
Yeah, this will make it possible to use it for coupled Euler with self-gravity :+1:

Reply:
Sweet!